### PR TITLE
Multimodality/vocab extension

### PR DESCRIPTION
**Pull request by nirmiger**
Vocab Extension for Mulitmodality

The previous vocabulary extension approach caused issues when working with larger models, particularly in setups using TP, PP, and FSDP.

This PR addresses those problems by moving the vocabulary extension logic from `checkpointing.py` to `loader_apertus.py`. As a result, the vocabulary is now extended in the Hugging Face (HF) format rather than the Megatron-LM format, which avoids the parallelism-related issues encountered previously.

This is a draft PR, as the feature has not yet been fully validated during training. So far, we have only confirmed that, at inference time in Hugging Face (HF) format, the model continues to produce the same logits for language tokens.

### CODE REVIEW

File: tools/checkpoint/loader_apertus.py [hunk 1]
```diff
@@ -88,6 +92,33 @@ def load_args_from_checkpoint(args, load_dir):
...
+    input rows (in the same dtype), so any forward pass that only indexes
+    token IDs ``< weight.shape[0]`` is bit-identical to the original.
+    """
+    old_vocab_size = weight.shape[0]
+    if new_vocab_size < old_vocab_size:
+        raise ValueError(
+            f"new_vocab_size={new_vocab_size} is smaller than current vocab size "
+            f"({old_vocab_size}); shrinking is not supported."
+        )
+    if new_vocab_size == old_vocab_size:
```
Comment (RaphaelKreft) [inline tools/checkpoint/loader_apertus.py:111]: Instead of a silent no-op we should be verbose about it. Ex. print info that no extension happened. Alternatively could raise an error.

File: tools/checkpoint/loader_apertus.py [hunk 2]
```diff
@@ -88,6 +92,33 @@ def load_args_from_checkpoint(args, load_dir):
...
+        return weight
+
+    n_new = new_vocab_size - old_vocab_size
+    w_f32 = weight.to(torch.float32)
+    mean = w_f32.mean(dim=0, keepdim=True)
+    std = w_f32.std(dim=0, keepdim=True)
+    new_rows = mean + std * torch.randn(n_new, w_f32.shape[1], dtype=torch.float32)
```
Comment (RaphaelKreft) [inline tools/checkpoint/loader_apertus.py:118]: For Apertus we want the extension to be reproducible. Make sure that the random generator is seeded before this torch.randn runs. In your test script on the other branch I saw you doing this manually.

1. In the test script `test_apertus_vocab_extension.py` extension is performed standalone on hf model and you manually set torch random seed.

2. If we plan to do it during conversion from hf -> megatron (using `tools/checkpoint/convert.py` with apertus loader) - we need to make sure that random seeding is added to be reproducible.

In my opinion the best option is to have a standalone tool on hf checkpoints. This makes it model agnostic and doesn't interfere with megatron. If we want to have the option on Megatron checkpoints as well, we should have this functionality for all different loaders.

File: tools/checkpoint/loader_apertus.py [hunk 3]
```diff
@@ -169,6 +200,36 @@ def _load_checkpoint(queue, args):
...
+        if new_vocab_size > old_vocab_size:
+            print(f"Extending vocab from {old_vocab_size} to {new_vocab_size} "
+                  f"(+{new_vocab_size - old_vocab_size} rows)")
+            word_embed = extend_embedding_table(word_embed, new_vocab_size)
+            if lm_head_weight is not None:
+                lm_head_weight = extend_embedding_table(lm_head_weight, new_vocab_size)
+
+            # Reflect the new vocab size in args/metadata so the saver builds a model
+            # of the right size and chunks the embedding correctly across TP.
+            true_vocab_size = new_vocab_size
```
Comment (RaphaelKreft) [inline tools/checkpoint/loader_apertus.py:224]: the extension of vocabulary in principle seems correct, however, its the users responisibility to make sure that the extended vocabulary size has enough padding so the total is divisible by the TP size. IMO we can improve here with different options:

1. We should in any case document the behavior of model extension more: explicitly state that the extension size is taken as is / or if we change that also respecting divisibility by the TP size
2. We could add the option to make the vocab size padded to be divisible by TP size (could be neat feature as current behavior is behaving the same way including padding).

**Review by RaphaelKreft — COMMENTED** [review]
Overall the logic looks good to me. Good work!

However there are a few design aspects to think about:

1. We should have the test script(`test_apertus_vocab_extension.py`) from the other branch(`nicola/multimodality/extend_vocab`) as part of the PR: it performs extension based on HF model checkpoints only which is versatile and good to have.
2. If we keep the extension in the Apertus Loader, so extension works during conversion with `tools/checkpoint/convert.py`, we should make this extension possible for different loaders, not only the apertus one. (ex. LLaMa/Mistral loader)
3. In any case, the behavior of model extension should be better documented, withing the extension script and/or the README.
4. In both cases we should also make sure that the random generator is seeded to make the extension reproducible. This is the case for the test script mentioned but not with `tools/checkpoint/convert.py`

Please look at the single comments for more details.

**Comment by RaphaelKreft** [comment]
Furthermore, the extension approach initializes the new rows with mean and std of the existing ones. This is a good approach, however we did it differently for the 8B model, where the Megatron init-std method was used. It would be good for final reproducibility if both methods would be supported.
